### DISCUSSION TITLE
Trouble with Schedule Entry Lock CC on Kwikset HomeConnect 620

### DISCUSSION THREAD

**Discussion by jasontduncan** [zwave-js/zwave-js#6441] [Q&A]
I'm super new to Zwave programming. I've spent days pouring through documentation, and experimenting with scripts in zwave-js, but I can't seem to make sense of the trouble I'm having with my Kiwkset Homeconnect 620. I'm trying to use the Schedule Entry Lock CC. The manual for the 620 says "Home Connect 620 smart lock supports multiple schedule types that can be assigned to access codes". When I query the 620 with 'supportsCC' and 'supportsCCAPI', it reports that Schedule Entry Lock CC is supported.  The trouble is that when I call 'getSchedules', 'numSlots', 'set*Schedule', etc, the lock reports there are 0 schedule slots. In the case of 'setYearDaySchedule', for example, the error states: "The schedule slot # must be between 1 and the number of supported day-of-year slots 0".  I can't find any way to create slots, on the lock. It makes no sense to me that Kwikset would build in support for the Schedule Entry Lock CC, but not have any slots to actually use it. I'm hoping this is just my noob-ness, and that a more experienced z-wave-er can point me in the right direction. Thank you to anyone who can provide some insight!!

Debug Log:
[zwavejs_2023-10-23.log](https://github.com/zwave-js/node-zwave-js/files/13074527/zwavejs_2023-10-23.log)

**Comment by AlCalzone** [comment]
Please make a [driver log](https://zwave-js.github.io/zwave-js-ui/#/troubleshooting/generating-logs?id=driver-logs), loglevel `debug` and attach it here as a file (drag & drop into the text field).

I'd like to see exactly what the device reports.

Also, do you have a manual for the device that actually explains the Z-Wave stuff like config parameters etc.? I can't find that online.

**Reply by jasontduncan** [reply]
I've added the log to my original message. It doesn't seem to include the output from my scripts, though.

**Reply by jasontduncan** [reply]
In regard to the zwave manual, I'm afraid not. All I found was the excerpt in my original message that came from the FAQ of the regular manual. I couldn't find any useful information from Kwikset :(

**Reply by AlCalzone** [reply]
```
2023-10-23T17:27:16.776Z DRIVER » [Node 002] [REQ] [SendDataBridge]
                                  │ source node id:   1
                                  │ transmit options: 0x25
                                  │ callback id:      3
                                  └─[Security2CCMessageEncapsulation]
                                    │ sequence number: 244
                                    │ extensions:
                                    │ · type: SPAN
                                    │   sender EI: 0x4b269fa6e48bff1765f2e4f8a4d287b2
                                    │ security class:  S2_AccessControl
                                    └─[ScheduleEntryLockCCSupportedGet]

...

2023-10-23T17:27:19.063Z CNTRLR   [Node 002] Timed out while waiting for a response from the node (ZW0201)
```

I think this is the problem. Z-Wave JS never learns about the number of supported schedules.

You may be able to figure out how many schedules the lock actually supports by circumventing the check in the CC API and sending the commands directly. Something like this should work for the weekday schedule - let me know if you need help adapting it for the other schedule types:

```ts
const { require } = this;
const {
  ScheduleEntryLockCCWeekDayScheduleGet,
  ScheduleEntryLockCCWeekDayScheduleSet,
  ScheduleEntryLockSetAction,
  ScheduleEntryLockWeekday,
} = require("@zwave-js/cc");

const nodeId = 2; // Put your node ID here
const endpoint = 0; // probably?

const userId = 1; // or whichever user ID this schedule is for
const slotId = 1; // increase this until the Get is no longer answered

const setWeekdaySchedule = new ScheduleEntryLockCCWeekDayScheduleSet(driver, {
  nodeId,
  endpoint,
  userId,
  slotId,
  action: ScheduleEntryLockSetAction.Set, // or .Erase to clear a schedule - then omit the day and tiem
  // Set whichever day and time you want
  weekday: ScheduleEntryLockWeekday.Sunday,
  startHour: 11,
  startMinute: 11,
  stopHour: 13,
  stopMinute: 13,
});

// Send the Set command
await driver.sendCommand(setWeekdaySchedule);

// Query it again to see if the change was accepted
const getWeekdaySchedule = new ScheduleEntryLockCCWeekDayScheduleGet(driver, {
  nodeId,
  endpoint,
  userId,
  slotId,
});

// Send the Get command
await driver.sendCommand(getWeekdaySchedule);

// Now watch the logs for the response
```

If you do find out how many schedules it supports, we can override the response for the command it doesn't respond to so this works out of the box.

**Reply by jasontduncan** [reply]
Thanks for the thoughtful response! Unfortunately, it looks like the same timeout is occurring on the get:
```
2023-10-24T19:39:32.391Z DRIVER   all queues busy
2023-10-24T19:39:32.393Z SERIAL » 0x0103003bc7                                                         (5 bytes)
2023-10-24T19:39:32.394Z DRIVER » [REQ] [GetBackgroundRSSI]
2023-10-24T19:39:32.398Z SERIAL « [ACK]                                                                   (0x06)
2023-10-24T19:39:32.398Z SERIAL « 0x0107013b9696969ace                                                 (9 bytes)
2023-10-24T19:39:32.399Z SERIAL » [ACK]                                                                   (0x06)
2023-10-24T19:39:32.401Z DRIVER « [RES] [GetBackgroundRSSI]
                                    channel 0: -106 dBm
                                    channel 1: -106 dBm
                                    channel 2: -106 dBm
2023-10-24T19:39:32.402Z DRIVER   all queues idle
2023-10-24T19:39:33.319Z DRIVER   all queues busy
2023-10-24T19:39:33.323Z SERIAL » 0x012400a900010002169f0305008d1bf522fb21533ff9fead6c5dd61c448eb3250 (38 bytes)
                                  00000001410
2023-10-24T19:39:33.325Z DRIVER » [Node 002] [REQ] [SendDataBridge]
                                  │ source node id:   1
                                  │ transmit options: 0x25
                                  │ callback id:      20
                                  └─[Security2CCMessageEncapsulation]
                                    │ sequence number: 5
                                    │ security class:  S2_AccessControl
                                    └─[ScheduleEntryLockCCWeekDayScheduleSet]
                                        user ID:    1
                                        slot #:     1
                                        action:     set
                                        weekday:    Sunday
                                        start time: 11:11
                                        end time:   13:13
2023-10-24T19:39:33.330Z SERIAL « [ACK]                                                                   (0x06)
2023-10-24T19:39:33.332Z SERIAL « 0x010401a90152                                                       (6 bytes)
2023-10-24T19:39:33.332Z SERIAL » [ACK]                                                                   (0x06)
2023-10-24T19:39:33.333Z DRIVER « [RES] [SendDataBridge]
                                    was sent: true
2023-10-24T19:39:34.570Z SERIAL « 0x011d00a91400007c00a47f7f7f7f01010300000000420100007f7f7f7f7fb8    (31 bytes)
2023-10-24T19:39:34.571Z SERIAL » [ACK]                                                                   (0x06)
2023-10-24T19:39:34.573Z DRIVER « [REQ] [SendDataBridge]
                                    callback id:            20
                                    transmit status:        OK, took 1240 ms
                                    routing attempts:       1
                                    protocol & route speed: Z-Wave, 40 kbit/s
                                    routing scheme:         LWR
                                    ACK RSSI:               -92 dBm
                                    ACK channel no.:        1
                                    TX channel no.:         1
                                    beam:                   1000 ms
2023-10-24T19:39:35.083Z SERIAL » 0x011e00a900010002109f030600f7f4e8353704d1a5934b3edf25000000001551  (32 bytes)
2023-10-24T19:39:35.084Z DRIVER » [Node 002] [REQ] [SendDataBridge]
                                  │ source node id:   1
                                  │ transmit options: 0x25
                                  │ callback id:      21
                                  └─[Security2CCMessageEncapsulation]
                                    │ sequence number: 6
                                    │ security class:  S2_AccessControl
                                    └─[ScheduleEntryLockCCWeekDayScheduleGet]
                                        user ID: 1
                                        slot #:  1
2023-10-24T19:39:35.089Z SERIAL « [ACK]                                                                   (0x06)
2023-10-24T19:39:35.091Z SERIAL « 0x010401a90152                                                       (6 bytes)
2023-10-24T19:39:35.091Z SERIAL » [ACK]                                                                   (0x06)
2023-10-24T19:39:35.092Z DRIVER « [RES] [SendDataBridge]
                                    was sent: true
2023-10-24T19:39:35.109Z SERIAL « 0x011d00a91500000100a47f7f7f7f01010300000000020100007f7f7f7f7f84    (31 bytes)
2023-10-24T19:39:35.109Z SERIAL » [ACK]                                                                   (0x06)
2023-10-24T19:39:35.109Z DRIVER « [REQ] [SendDataBridge]
                                    callback id:            21
                                    transmit status:        OK, took 10 ms
                                    routing attempts:       1
                                    protocol & route speed: Z-Wave, 40 kbit/s
                                    routing scheme:         LWR
                                    ACK RSSI:               -92 dBm
                                    ACK channel no.:        1
                                    TX channel no.:         1
2023-10-24T19:39:36.139Z CNTRLR   [Node 002] Timed out while waiting for a response from the node (ZW0201)
2023-10-24T19:39:36.141Z DRIVER   all queues idle
``` 
I tried user 1's code on the lock, just in case the set is working but not the get. Sadly, the code still works.
EDIT: Scratch that. I was using the code for a different user. The user 1 code no longer works, which is a start!

**Reply by AlCalzone** [reply]
Not being able to query schedules (even if they work) is going to be a headache though.
A few more notes: your lock may also just support YearDay schedules or DailyRepeating schedules. It may also interpret the fields wrongly - I have a lock that interprets the duration of DailyRepeating schedules as an end time instead.

**Reply by jasontduncan** [reply]
Agreed. I'm not even certain that the schedule is working, just that the door code stops working when I set the schedule. Since I'm unable to get the dateTime from the lock, I have no idea what the lock thinks the current date and time are, so I can't set a schedule to test. I was able to set and get the dateAndTime, on the TimeParametersCC, but that doesn't seem to be related (and I haven't come across any documentation that clarifies what TimeParametersCC is even for).
I really appreciate your input on this. You've given me enough new understanding about the system that I don't feel so stuck. I'll hack away at it, and see what I can figure out.
Thanks again!

**Reply by AlCalzone** [reply]
Time parameters CC is used to set the time and date on a device. You can't set the timezone though, this needs to be done with aa different CC.
There's a helper method though that does the right thing(TM), and AFAIK this is accessible through the UI.

**Comment by jasontduncan** [comment]
Thanks, I’ll take a look for it in the UI. Something I noticed about  Time
Parameters CC, though, is that the time never changes, when I call get. It
seems to behave more like a timestamp. I’m not one to suspect incompetence
in anyone but myself, but is it possible Kwikset botched the Z-wave
implementation?

**Reply by AlCalzone** [reply]
Seems they botched quite a lot in that lock: https://github.com/zwave-js/node-zwave-js/discussions/5566

**Reply by jasontduncan** [reply]
That's a bummer. When I try to find information about the best locks to use with Z-wave, all I can find is marketing bullshit. Do you happen to know of (maybe you can point me toward a list) any Z-wave locks that behave better? I bought an Assure Lock 2 with Z-wave, the same time as the Kwikset, but it arrived without it's Z-wave module, and I'm still waiting for Yale to respond to my request for the part. Maybe it would work better (assuming they ever send me a module)?
